cstore_reader.c

```diff
@@ -276,16 +382,76 @@ CStoreReadNextRow(TableReadState *readState, Datum *columnValues, bool *columnNu
 void
 CStoreEndRead(TableReadState *readState)
 {
+	int columnCount = readState->tupleDescriptor->natts;
+
 	MemoryContextDelete(readState->stripeReadContext);
 	FreeFile(readState->tableFile);
 	list_free_deep(readState->tableFooter->stripeMetadataList);
+	FreeColumnBlockDataBuffers(readState->blockDataBuffersArray, columnCount);
 	pfree(readState->tableFooter);
 	pfree(readState);
 }
 
 
 /*
- * LoadFilteredStripeData reads and decompresses stripe data from the given file.
+ * CreateBlockDataBuffersArray creates data buffers to keep deserialized exist and
+ * value arrays for requested columns in columnMask.
+ */
+BlockDataBuffers **
+CreateBlockDataBuffersArray(uint32 columnCount, bool* columnMask, uint32 blockRowCount)
+{
+	uint32 columnIndex = 0;
+	BlockDataBuffers **blockDataBuffersArray =
+			palloc0(columnCount * sizeof(BlockDataBuffers*));
+
+	/* allocate block memory for deserialized data */
+	for (columnIndex = 0; columnIndex < columnCount; columnIndex++)
+	{
+		if (columnMask[columnIndex])
+		{
+			BlockDataBuffers *blockDataBuffers = palloc0(sizeof(BlockDataBuffers));
+
+			blockDataBuffers->existsArray = palloc0(blockRowCount * sizeof(bool));
+			blockDataBuffers->valueArray = palloc0(blockRowCount * sizeof(Datum));
+			blockDataBuffers->serializedValueBuffer = NULL;
+			blockDataBuffersArray[columnIndex] = blockDataBuffers;
+		}
+	}
+	
+	return blockDataBuffersArray;
+}
+
+
+/*
+ * FreeColumnBlockDataBuffers creates data buffers to keep deserialized exist and
+ * value arrays for requested columns in columnMask.
+ * BlockDataBuffers->serializedValueBuffer lives in memory read/write context
+ * so it is deallocated automatically when the context is deleted.
+ */
+void
+FreeColumnBlockDataBuffers(BlockDataBuffers **blockDataBuffersArray,
+						   uint32 columnCount)
+{
+	uint32 columnIndex = 0;
+
+	for (columnIndex = 0; columnIndex < columnCount; columnIndex++)
+	{
+		BlockDataBuffers *blockDataBuffers = blockDataBuffersArray[columnIndex];
+
```

Comment:
In CitusDB's source code, when we do ```X x = ...```, we usually don't separate it from ```if (x ...)``` with newline. i.e. we do something like:

```C
BlockDataBuffers *blockDataBuffers = blockDataBuffersArray[columnIndex];
if (blockDataBuffers != NULL)
{
 ...
}
```

instead of

```C
BlockDataBuffers *blockDataBuffers = blockDataBuffersArray[columnIndex];

if (blockDataBuffers != NULL)
{
 ...
}
```

I don't the reason why we're doing this always. I don't know if there is much value in being consistent in these little details.